### ISSUE TITLE
OpenGL Performance issue and Filters 

### ISSUE THREAD

**Issue by hehk1234** [rafaelvcaetano/melonDS-android#1275]
Good morning Rafael, what's up?

I saw that support for the OGL API was recently added, however when we upscale more aggressively, for example, 5,6,7 and 8x the performance drops dramatically (Snapdragon 8 gen 2 - AYA Odin 2 Pro). I think it's possible to use 8x with ease on this processor (not that I'm going to, 1080P is excellent for my screen, but just giving feedback that there must be some bottleneck preventing this from happening in the code).

And I would like to check the availability of implementing the filters again together with OGL. 

Thank you very much for your attention!

**Comment by eebssk1** [comment]
On my device,the audio becomes super laggy with opengl mode defaut 1x.

**Comment by eebssk1** [comment]
> I'm using a Moto G54 (mid range specs), performance drops heavily when I switch to OpenGL (even when the resolution is set to 1x), I'm not sure why this is the case for it runs mostly fine on software.

Looks like device driver problem. I have similar issues on PPSSPP, OpenGL mode is laggy and unsmooth compared to vulkan.

**Comment by ElTioRata** [comment]
Vulkan works on PPSSPP in my case but sometimes it crashes. PPSSPP's OpenGL is stable and I haven't experienced any performance issues.

**Comment by hehk1234** [comment]
@rafaelvcaetano How are you, my friend? I've been taking a look at this part of OpenGL, and I've noticed that there seems to be some part of the code that isn't “talking” to the kernel correctly. When I locked the Odin 2 clock at 400mhz on the GPU, the 10x upscale worked very well, but on my Galaxy S24U the 10x upscale does not work well, when I used the scene to check, the GPU clock was at 232mhz and to try to force it, I recorded the screen and automatically the clock went up to 450mhz and the 10x upscale worked perfectly, but when I stopped recording the screen, the clock goes back to 232mhz and the emulation is choked. Perhaps an adjustment is needed in the code so that OGL correctly requests the power of the hardware as required.

**Comment by rafaelvcaetano** [comment]
Yeah, the current implementation is not perfect. Currently, when the frame finishes, the CPU was to wait for the GPU to finish drawing so that it can copy the pixels to draw to the screen. The correct approach would be to let the GPU directly draw to the screen, but there are considerable changes that I need to make to allow that to happen and I'm still thinking of the best way to do it.

**Comment by hehk1234** [comment]
// FBO Initialization
GLuint fbo, texture;
glGenFramebuffers(1, &fbo);
glGenTextures(1, &texture);

// Configure the texture to store FBO data
glBindTexture(GL_TEXTURE_2D, texture);
glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, screenWidth, screenHeight, 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

// Attach the texture to the FBO
glBindFramebuffer(GL_FRAMEBUFFER, fbo);
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, texture, 0);

// Check if the FBO is set up correctly
if (glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE) {
    // Handle error
}

// Render to FBO
glBindFramebuffer(GL_FRAMEBUFFER, fbo);
glViewport(0, 0, screenWidth, screenHeight);

// Perform regular rendering here (calls like glDrawArrays/glDrawElements)

// Final Step - Render the FBO directly to the screen
glBindFramebuffer(GL_FRAMEBUFFER, 0);
glViewport(0, 0, screenWidth, screenHeight);

// Render a fullscreen quad using the FBO texture
glEnable(GL_TEXTURE_2D);
glBindTexture(GL_TEXTURE_2D, texture);

glBegin(GL_QUADS);
glTexCoord2f(0.0f, 0.0f); glVertex2f(-1.0f, -1.0f);
glTexCoord2f(1.0f, 0.0f); glVertex2f(1.0f, -1.0f);
glTexCoord2f(1.0f, 1.0f); glVertex2f(1.0f, 1.0f);
glTexCoord2f(0.0f, 1.0f); glVertex2f(-1.0f, 1.0f);
glEnd();

**Comment by rafaelvcaetano** [comment]
Yeah, it's something like that. But the problem here is not knowing how to do it. The hard part is deciding on the best approach to implement that on top of the existing stuff. If I choose the wrong approach, performance will suffer and changing it won't be easy.

**Comment by hehk1234** [comment]
> Yeah, it's something like that. But the problem here is not knowing how to do it. The hard part is deciding on the best approach to implement that on top of the existing stuff. If I choose the wrong approach, performance will suffer and changing it won't be easy.

I understood. I had opened a PR about resizing the screen, but the adjustment didn't turn out quite the way I'd imagined. As a reference, you have citra mmj, which resizes in a “free” way, for example, I can adjust with absurd precision, occupying both screens proportionally and equally on the device, without any black space left over. If it's possible to look into this in the future, I'd be grateful, it would be great

**Comment by Fatcat2024-cpu** [comment]
Try the last nightly build https://github.com/rafaelvcaetano/melonDS-android/releases

**Comment by ElTioRata** [comment]
Needs some more testing but now it performs well.
